services/device-registry/src/test/java/org/eclipse/hono/deviceregistry/DeviceRegistryRestServerTest.java

```diff
@@ -0,0 +1,430 @@
+@RunWith(VertxUnitRunner.class)
+public class DeviceRegistryRestServerTest {
+
+    private static final String HOST = "localhost";
+
+    private static final String TENANT = "testTenant";
+    private static final String DEVICE_ID = "testDeviceId";
+
+    private static Vertx vertx;
+    private static FileBasedRegistrationService registrationService;
+    private static DeviceRegistryRestServer deviceRegistryRestServer;
+
+    @BeforeClass
+    public static void setUp(final TestContext context) throws Exception {
+        vertx = Vertx.vertx();
+
+        Future<String> setupTracker = Future.future();
+        setupTracker.setHandler(context.asyncAssertSuccess());
+
+        ServiceConfigProperties restServerProps = new ServiceConfigProperties();
+        restServerProps.setInsecurePortEnabled(true);
+        restServerProps.setInsecurePort(0);
+
+        RegistrationHttpEndpoint registrationHttpEndpoint = new RegistrationHttpEndpoint(vertx);
+        deviceRegistryRestServer = new DeviceRegistryRestServer();
+        deviceRegistryRestServer.addEndpoint(registrationHttpEndpoint);
+        deviceRegistryRestServer.setConfig(restServerProps);
+
+        FileBasedRegistrationConfigProperties regServiceProps = new FileBasedRegistrationConfigProperties();
+        registrationService = new FileBasedRegistrationService();
+        registrationService.setConfig(regServiceProps);
+        SignatureSupportingConfigProperties signatureSupportingConfProps = new SignatureSupportingConfigProperties();
+        signatureSupportingConfProps.setSharedSecret("DeviceRegistrySharedSecret_HasToBe32CharsOrLonger");
+        registrationService.setRegistrationAssertionFactory(
+                RegistrationAssertionHelperImpl.forSigning(vertx, signatureSupportingConfProps));
+
+        Future<String> restServerDeploymentTracker = Future.future();
+        vertx.deployVerticle(deviceRegistryRestServer, restServerDeploymentTracker.completer());
+        restServerDeploymentTracker.compose(s -> {
+            Future<String> registrationServiceDeploymentTracker = Future.future();
+            vertx.deployVerticle(registrationService, registrationServiceDeploymentTracker.completer());
+            return registrationServiceDeploymentTracker;
+        }).compose(c -> setupTracker.complete(), setupTracker);
+    }
+
+    @AfterClass
+    public static void tearDown(final TestContext context) throws Exception {
+        vertx.close(context.asyncAssertSuccess());
+    }
+
+    @After
+    public void clearRegistry() throws InterruptedException {
+        registrationService.clear();
+    }
+
+    private int getPort() {
+        return deviceRegistryRestServer.getInsecurePort();
+    }
+
+    @Test
+    public void testAddDevice(final TestContext context) throws Exception {
+        final String requestUri = String.format("/%s/%s", REGISTRATION_ENDPOINT, TENANT);
+        final String requestBody = "{\"" + FIELD_DEVICE_ID + "\":\"" + DEVICE_ID + "\",\"test\":\"test\"}";
```

Comment:
A general proposal:
to improve readability I would rather use String.format.

E.g. like the following:
`
final String requestBody2 = String.format("{\"%s\":\"%s\",\"test\":\"test\"}", FIELD_DEVICE_ID, DEVICE_ID);
`

or (even more readable):
`
final String requestBody3 = String.format("{'%s':'%s','test':'test'}", FIELD_DEVICE_ID, DEVICE_ID).replace('\'','"');
`

WDYT?

Reply:
Yes, that string concatenating/quoting better be improved.
I like your 2nd solution. In this scenario of creating a JSON string, there's also the possibility to use `JsonObject`:
`new JsonObject().put(FIELD_DEVICE_ID, DEVICE_ID).put("test", "test").encode();`.
I think that's even more readable.

Reply:
:+1: That is indeed even better!

Reply:
You should use `JsonObject.encodePrettily()`, though. Otherwise the whole JSON be on one line ...

Reply:
But I think that isn't an issue here. For the cases here, the JSON is short, containing only 2-3 key-value pairs, so during debugging the string value would still be readable.